### PR TITLE
setup lobby creation for multiplayer

### PR DESCRIPTION
**Pull request by dhillonJas**
Frontend and backend
- Initial lobby creation setup
- Front end change to show room id
- Front end change to make main menu look a little better and add options to create and join room
- Shows number of players in a lobby as cards above table area

### CODE REVIEW

File: digital-deck-ui/src/socket.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import io from 'socket.io-client'
+const socket = io('http://localhost:5000')
```
Comment (mdangnguyen) [inline digital-deck-ui/src/socket.js:2]: We don't want to just immediately connect the user to the server the moment they start up the app. We should only establish the connection when a user creates/joins a room.

File: digital-deck-ui/src/components/MainMenu.vue [hunk 2]
```diff
@@ -96,35 +151,27 @@ export default {
             this.isLoaded = true;
           });
       }
+      this.$socket.emit('joinRoom', this.$store.getters.getSessionId);
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:154]: Putting this outside the `if` will have the player join the room regardless of whether the session was successfully initialized. This may also be related to a bug I'm seeing where after clicking `Create` the page just reloads and the url becomes `http://localhost:3000/?`

File: digital-deck-ui/src/components/MainMenu.vue [hunk 3]
```diff
@@ -1,15 +1,57 @@
 <template>
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:1]: We'll want another page for the lobby. It should show up after a player creates/joins the room. It's like a waiting room so other players can join before the game begins. The host will have the option to start the game whenever they want and then all the players will be moved to the game screen.

File: digital-deck-server/index.js [hunk 4]
```diff
@@ -5,8 +5,12 @@ const http = require("http");
 const { Server } = require('socket.io');
 const sessionRoute = require('./routes/session');
 const playerRoute = require('./routes/player');
+const {handleSocket} = require('./socket');
 
 const app = express();
+
+app.use(express.static('../digital-deck-ui'));
```
Comment (mdangnguyen) [inline digital-deck-server/index.js:12]: What does this do?

**Comment by mdangnguyen** [comment]
We also need a playerId generation logic since what we have right now doesn't work for multiplayer. You can work on that for this ticket or it can be a part of the join lobby ticket, either way's fine.

File: digital-deck-ui/src/components/MainMenu.vue [hunk 5]
```diff
@@ -1,15 +1,57 @@
...
+              required=""
+            />
+            <label>Cards on hand</label>
+          </div>
+
+          <div class="number-box">
+            <input
+              v-model.number="cardsOnTable"
+              type="number"
+              min="1"
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:36]: The `min` value for `cardsPerPlayer` and `cardsOnTable` here should be 0. You can also make it so that the default value for `deckSelected` is 1 if you want.

File: digital-deck-ui/src/components/MainMenu.vue [hunk 6]
```diff
@@ -1,15 +1,57 @@
...
-      <span>Decks</span><br />
-      <input v-model.number="deckSelected" type="number" min="1" oninput="validity.valid||(value='');"/><br />
-      <span>Cards on hand</span><br />
-      <input v-model.number="cardsPerPlayer" type="number"  min="1" oninput="validity.valid||(value='');"/><br />
-      <span>Cards on table</span><br />
-      <input v-model.number="cardsOnTable" type="number"  min="0" oninput="validity.valid||(value='');"/><br /><br />
-      <button @click="submitForm" style="cursor: pointer;">Start</button>
+      <div v-if="form1" class="input-form">
+        <form>
+          <div class="number-box">
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:10]: Can you please also add another field so players can input their name and store this data in the backend as well? We'll need that for player display

File: digital-deck-ui/src/components/OtherPlayers.vue [hunk 7]
```diff
@@ -0,0 +1,46 @@
+<template>
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/OtherPlayers.vue:1]: Can you rename this to something like `PlayerList` please? `OtherPlayers` sound weird.

File: digital-deck-ui/src/components/MainMenu.vue [hunk 8]
```diff
@@ -1,15 +1,57 @@
...
+            <input
+              v-model.number="cardsOnTable"
+              type="number"
+              min="1"
+              oninput="validity.valid||(value='');"
+              required=""
+            />
+            <label>Cards on table</label>
+          </div>
+
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:42]: If you want, we can also add another field to input the room's password. It's not a priority or anything, but it's something you can work on if you want.

File: digital-deck-ui/src/components/MainMenu.vue [hunk 9]
```diff
@@ -96,35 +151,27 @@ export default {
...
       }
+      this.$socket.emit('joinRoom', this.$store.getters.getSessionId);
     },
 
     closeSession() {
       $cookies.set("SessionId", -1, "1h");
       this.isLoaded = false;
     },
+
+    toggleForm1() {
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:162]: Can you rename these 2 functions to something like `ToggleCreateForm` and `ToggleJoinForm`?

File: digital-deck-ui/src/components/MainMenu.vue [hunk 10]
```diff
@@ -42,9 +84,12 @@ export default {
   data: () => {
     return {
       isLoaded: false,
-      deckSelected: '',
-      cardsPerPlayer: '',
-      cardsOnTable: '',
+      deckSelected: "",
+      cardsPerPlayer: "",
+      cardsOnTable: "",
+      form1: false,
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:90]: Can you rename `form1` and `form2` to something more descriptive?

File: digital-deck-ui/src/components/MainMenu.vue [hunk 11]
```diff
@@ -42,9 +84,12 @@ export default {
...
+      form2: false,
+      roomJoined: 0,
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:92]: I don't think this is being used anywhere

File: digital-deck-ui/src/main.js [hunk 12]
```diff
@@ -1,7 +1,12 @@
-import { createApp } from 'vue';
-import App from './App.vue';
-import store from './store';
+import { createApp } from 'vue'
```
Comment (mdangnguyen) [inline digital-deck-ui/src/main.js:1]: Can you put the `;` after every line please? I know it's not required for javascript but it's good practice

File: digital-deck-ui/src/store/index.js [hunk 13]
```diff
@@ -53,6 +58,7 @@ export default createStore({
                 commit('setPlayerCards', res.data.players[0].cards);
                 commit('setTableCards', res.data.table);
                 commit('setCardsInDeck', res.data.deck);
+                commit('setNumPlayers', res.data.numPlayers);
```
Comment (mdangnguyen) [inline digital-deck-ui/src/store/index.js:61]: Instead of `numPlayers`, I think it would be more practical if we keep an array of player names instead.

File: digital-deck-ui/src/socket.js (hunk 1 above)
Comment (dhillonJas) [reply digital-deck-ui/src/socket.js:2]: I think this is fine for now. This also lets us to access this.socket all over the client side, if it is created in a component then we will have to either create connection for each component or pass the socket somehow that makes the code a little 
messier.

File: digital-deck-ui/src/components/MainMenu.vue (hunk 2 above)
Comment (dhillonJas) [reply digital-deck-ui/src/components/MainMenu.vue:154]: yeah, I changed it to only join when form is correct but I still the `?` in the url. I think this may be related to cookies

File: digital-deck-ui/src/components/MainMenu.vue (hunk 3 above)
Comment (dhillonJas) [reply digital-deck-ui/src/components/MainMenu.vue:1]: Okay, I created this page but this is gonna show the same component that we show in Game

File: digital-deck-server/index.js (hunk 4 above)
Comment (dhillonJas) [reply digital-deck-server/index.js:12]: it makes it possible to access files from this folder via HTTP. It was done in all socket tutorials, I don't know if we require it or not

File: digital-deck-ui/src/components/MainMenu.vue (hunk 6 above)
Comment (dhillonJas) [reply digital-deck-ui/src/components/MainMenu.vue:10]: That is gonna required a lot of change in the backend. It will be a big change as some of our tests will also need to be changed. And if you look in `CardDistService`, I don't see how we can get the names there without changing a lot of things. I think we can do without names for now, as it is not really a core feature.

File: digital-deck-ui/src/store/index.js (hunk 13 above)
Comment (dhillonJas) [reply digital-deck-ui/src/store/index.js:61]: This again will require a lot of change when handling requests and sending data, which in turn may make our tests fail.

File: digital-deck-ui/src/components/MainMenu.vue (hunk 6 above)
Comment (mdangnguyen) [reply digital-deck-ui/src/components/MainMenu.vue:10]: The way I see it, a lot of changes have to be made to `CardDistService` regardless, since I did point out early in our first sprint that the current code will not work for multiplayer. Aside from that, I don't think adding names will be that much work, since it's just 1 more instance variable in `Player.js`. I also think it's a core part of acceptance test since I don't think our app would make sense if none of the players can tell who is who.

File: digital-deck-ui/src/components/MainMenu.vue (hunk 2 above)
Comment (mdangnguyen) [reply digital-deck-ui/src/components/MainMenu.vue:152]: You should put this inside `finally`, though you should probably change it into `then` since everything in `finally` runs regardless of whether there was an error or not, and we don't want to continue if there was an error.

File: digital-deck-ui/src/socket.js (hunk 1 above)
Comment (mdangnguyen) [reply digital-deck-ui/src/socket.js:2]: Yeah, actually this is fine. I was concerned that there would be multiple socket connections if the the client refreshes the page or something, but apparently socket.io automatically disconnects in that case so there's no problem there.

File: digital-deck-ui/src/components/MainMenu.vue (hunk 3 above)
Comment (mdangnguyen) [reply digital-deck-ui/src/components/MainMenu.vue:1]: How do you mean?

File: digital-deck-ui/src/store/index.js (hunk 13 above)
Comment (mdangnguyen) [reply digital-deck-ui/src/store/index.js:61]: We'll have to modify the tests in that case, since they are not useful for the current state of our app.

File: digital-deck-ui/src/components/PlayersList.vue [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+<script>
+import CardDeckImageEnum from "./CardDeckImageEnum.js";
+
+export default {
+  name: "players-list",
+  computed:{
+    // filters this player
+    playersList : function(){
+      return this.$store.getters.getPlayersInfo.filter(player => player.name != this.$store.getters.getName)
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/PlayersList.vue:26]: Maybe you can add a flag here like `return isLobby ? this.$store.getters.getPlayersInfo : this.$store.getters.getPlayersInfo.filter(player => player.name != this.$store.getters.getName)` and have `isLobby` be a prop since it makes sense to display all the players in the lobby.

Comment (mdangnguyen) [reply digital-deck-ui/src/components/PlayersList.vue:26]: We probably also don't want to show the number of cards for each player while they're in the lobby

Comment (mdangnguyen) [reply digital-deck-ui/src/components/PlayersList.vue:26]: Nvm, I just noticed you don't use this component in `Lobby.vue`. You probably don't have to commit this file then since Surya will be working on it

File: digital-deck-ui/src/components/Lobby.vue [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+      </div>
+    </div>
+
+    <game v-if="isLoaded" />
+  </div>
+</template>
+
+<script>
+import Game from "./Game.vue";
+import PlayersList from "./PlayersList.vue";
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/Lobby.vue:37]: `PlayersList` isn't used here

File: digital-deck-ui/src/components/MainMenu.vue [hunk 16]
```diff
@@ -89,42 +152,36 @@ export default {
         this.$store
           .dispatch("initSession", {
             decks: this.deckSelected,
+            name: this.name,
             cardsPerPlayer: this.cardsPerPlayer,
             cardsOnTable: this.cardsOnTable,
           })
-          .finally(() => {
+          .then(() => {
+            // this.$socket.emit('joinRoom', this.$store.getters.getSessionId);
```
Comment (mdangnguyen) [inline digital-deck-ui/src/components/MainMenu.vue:160]: Pls make sure to fix the issues related to `joinRoom` before merging

**Review by mdangnguyen — APPROVED** [review]
